### ISSUE TITLE
Stress rig waveform generator hydraulics tripping

### ISSUE THREAD

**Issue by Tom-Willemsen** [ISISComputingGroup/IBEX#2802]
> The waveform generator option for the Engin-X Instron stress rig in IBEX is 
> faulty. It will run the waveform, but after (usually) an 8-second timeout or 
> so, the hydraulics will switch off with an error code of 20 from the Q415 GPIB 
> query. After resetting the hydraulics, even ordinary move commands can 
> subsequently trigger the same fault, with the 'remote' light on the hardware 
> control panel sometimes remaining lit after the move command, which should 
> normally never happen. A full hardware power cycle of the rig was required to 
> restore orinary behaviour
> 
> A fault like what is seen here with the waveform generator was apparently also 
> seen before when first testing the IBEX stressrig interface with ordinary move 
> commands, a bug which has since been fixed. Since the ordinary move commands 
> are actually implemented as waveforms on the Instron hardware, it is possible 
> that the bug fix employed before merely needs to be extended to the waveform 
> generator.
> 
> For the time being, the affected experiment is running with the older SECI 
> interface.
> 
> Reported by: Joe Kelleher (mailto:joe.kelleher@stfc.ac.uk)
> ISIS Instrument: ENGINX
> Name of computer problem was submitted from: ndxenginx.isis.cclrc.ac.uk
> Current ISIS Cycle: 17/3
> Product: Instrument Problem/Bug Report
> Category: Computing: Sample environment/Device Control
> Severity: major
> Priority: P3

Last time we had a similar issue, it was fixed by disabling the watchdog before every command. That fix was applied generally across all commands.

I believe this is the stress rig's response to "I got a command which was valid but I don't want it right now", so it could also be a timing thing. Perhaps the waveform start sequence is sending commands either too slowly, too fast, or in the wrong order?

As a potential debugging step, I would suggest commenting out the `P909,0` command in the stress rig protocol and seeing if that helps anything. If it does, it is possibly watchdog related. If not, it is possibly timing/sequencing related.

**Comment by ChrisM-S** [comment]
Could be that the stress rig has already produced a response to a previous command that needs reading to keep reads/writes in sync.

**Comment by Tom-Willemsen** [comment]
Can't be done while ENGIN-X are using their beamline (i.e. within cycle). Needs to wait until probably January.

**Comment by AdrianPotter** [comment]
Emailed Joe to see if there's a good time

**Comment by Tom-Willemsen** [comment]
Saurabh just popped into the office. He will try to set up the rig for us Thursday morning. Once he's set it up he's happy for us to fault-find whenever.

**Comment by Tom-Willemsen** [comment]
The scientists have "set up the rig" for us in that we can talk to it, but the actuator is not switched on so I can't run a waveform (which is the whole point of this ticket). It's not immediately clear how to switch on the actuator. I have emailed the scientists and will chase.

**Comment by Tom-Willemsen** [comment]
To turn the hydraulics/actuator on for the 50kN rig you have to hold the "hydraulics on" button down for 10 seconds. I'd only been trying pressing it briefly (in which case it trips off immediately).

**Comment by Tom-Willemsen** [comment]
Impeded until the south side cooling water is turned back on.

**Comment by Tom-Willemsen** [comment]
Meeting with Joe on Engin-X at 1:45 tomorrow to test the waveform in stress and strain control modes.

**Comment by Tom-Willemsen** [comment]
This appeared to be caused by a few different things:
- Waveform generator was sometimes being stopped while starting. Fixed in https://github.com/ISISComputingGroup/EPICS-ioc/pull/223
- Device (very very occasionally) timed out. Fixed in https://github.com/ISISComputingGroup/EPICS-instron/pull/5
- Instrument scripts writing waveform setpoints to the device too fast. I added some 0.5 second delays to their instrument scripts which seemed to dramatically increase stability
- I have added documentation at https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Instron-stress-rig under the gotchas section
